### ISSUE TITLE
Mentioning how to retrieve the public key

### ISSUE THREAD

**Issue by cwebber** [w3c-dvcg/ld-signatures#4]
[Section 7](https://w3c-dvcg.github.io/ld-signatures/#signature-verification-algorithm) mentions retrieval of the public key, and obviously that's a pretty key part of this process.

I've expressed [my impressions](https://github.com/w3c/activitypub/issues/203#issuecomment-297553229) of how to retrieve the public key; are they correct?  It would be good to have this written up in the document.

**Comment by msporny** [comment]
Yes, we need to do this. The algorithm is basically this:


1. You typically start out with a signed document and you get to the Entity Profile document (aka "Person") by checking the ```creator``` field in the signature, which is typically a URL that you can dereference.
2. You dereference the ```creator``` URL and get an Entity Profile document (aka "Person"). That document MUST contain at least the public key ID in the list of public keys owned by the Entity Profile.
3. The Entity Profile document (aka "Person") MUST be provided over a secure communications channel (such as TLS)... or be digitally signed in some way. If it's not signed, you're depending on TLS and the Certificate Authority system to ensure a valid document. If it's signed, you're depending on whatever key management system you're using to ensure a valid document. Doing both is recommended, but not required.
4. The Entity Profile document (aka "Person") MUST reference the public key identifier and SHOULD include the public key PEM.
5. The Key must point back to its owner.

We need to get this into the document in a more generic way. I'll try to add in the next week or so.

**Comment by dlongley** [comment]
Minor fixes to the algorithm:

1. You start out with a signed document. The signature node contains a the public key identifier in the `creator` field, typically as a URL. You can dereference this URL to get the document containing the public key information. If it's not a URL, do some other scheme-specific operation to get the document that contains the public key information.
2. The public key information should contain at least a reference to the `owner` of that public key (e.g. the URL identifier for an Entity Profile document (often aka "Person", but could be a "Thing"). You then deference the URL identifier for the Entity Profile document. Note: If the public key ID is a fragment-based URL (e.g. `https://foo.com/me#my-key`), it could be that the Entity Profile document and the public key document are one in the same.
3. The Entity Profile document MUST contain at least the public key ID in the list of public keys owned by the Entity Profile (e.g. `{"@context": "...", "publicKey": [<id in here>]}`). The Entity Profile document MUST also be provided over a secure communications channel (such as TLS) or be digitally signed in some way. If it's not signed, you're depending on TLS and the Certificate Authority system to ensure a valid document. If it's signed, you're depending on whatever key management system you're using to ensure a valid document. Doing both is recommended, but not required.
4. The Entity Profile document MUST reference the public key identifier and the public key document or the Entity Profile document MUST include the public key PEM.

In short, you must dereference the key to get its `owner` and PEM. The key's `owner` must be an Entity Profile's ID (the entity you are confirming as the author of the signature) and deferencing the Entity Profile ID should yield a document that declares that the Entity Profile owns key (by listing its ID) confirming a "bi-directional" relationship.

**Comment by cwebber** [comment]
So one tricky thing that's probably worth including is: when you retrieve the public key, what do you set as the `Accept` header?  I would think `application/ld+json` would be the default?

I know we also want to support triple representations that aren't necessarily json-ld.  In that case, maybe the right option is to provide an `options['getPublicKey']` option which is a method with a lot more leeway as to how it retrieves the key (I suppose it would als oneed to do the work of verifying the owner, etc?).  If we did that though, we'd still want to document how the default method operates.

**Comment by dlongley** [comment]
@cwebber,

>  In that case, maybe the right option is to provide an options['getPublicKey'] option which is a method with a lot more leeway as to how it retrieves the key

^That sounds a lot like this concept:

https://github.com/digitalbazaar/jsonld-signatures/pull/23#issuecomment-308193545
